File: Cpp/FirstAndLastElement.cpp
#include <iostream>
#include <vector>
using namespace std;

class FirstAndLastElement {
public:
    static vector<int> searchRange(vector<int>& nums, int target) {
        int first = findFirstOccurance(nums, target, 0, nums.size() - 1);
        int last = findLastOccurance(nums, target, 0, nums.size() - 1);
        
        return vector<int>{first, last};
    }
    
    static int findFirstOccurance(vector<int>& nums, int target, int start, int end) {
        int index = -1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(nums[mid] == target) {
                end = mid - 1;
                index = mid;
            } else if(nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        
        return index;
    }
    
    static int findLastOccurance(vector<int>& nums, int target, int start, int end) {
        int index = -1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(nums[mid] == target) {
                start = mid + 1;
                index = mid;
            }else if(nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        
        return index;
    }
};

int main() {
    vector<int> nums = {5,7,7,8,8,10};
    vector<int> res = FirstAndLastElement::searchRange(nums, 8);
    cout << "{" << res[0] << " " << res[1] << "}" << endl;
    res = FirstAndLastElement::searchRange(nums, 6);
    cout << "{" << res[0] << " " << res[1] << "}" << endl;
    nums = {};
    res = FirstAndLastElement::searchRange({nums}, 0);
    cout << "{" << res[0] << " " << res[1] << "}" << endl;
    return 0;
}

File: Java/FirstAndLastElement.java
package Java;

class FirstAndLastElement {
    public static int[] searchRange(int[] nums, int target) {
        int first = findFirstOccurance(nums, target, 0, nums.length - 1);
        int last = findLastOccurance(nums, target, 0, nums.length - 1);
        
        return new int[]{first, last};
    }
    
    public static int findFirstOccurance(int[] nums, int target, int start, int end) {
        int index = -1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(nums[mid] == target) {
                end = mid - 1;
                index = mid;
            } else if(nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        
        return index;
    }
    
    public static int findLastOccurance(int[] nums, int target, int start, int end) {
        int index = -1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(nums[mid] == target) {
                start = mid + 1;
                index = mid;
            }else if(nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        
        return index;
    }

    public static void main(String[] args) {
        int[] res = searchRange(new int[] {5,7,7,8,8,10}, 8);
        System.out.println("{"+res[0] +" " + res[1]+"}");
        res = searchRange(new int[] {5,7,7,8,8,10}, 6);
        System.out.println("{"+res[0] +" " + res[1]+"}");
        res = searchRange(new int[] {}, 0);
        System.out.println("{"+res[0] +" " + res[1]+"}");
    }
}
